src/KafkaW/Producer.cpp

```diff
@@ -1,171 +1,82 @@
+Producer::Producer(BrokerSettings &Settings)
     : ProducerBrokerSettings(Settings) {
-  id = g_kafka_producer_instance_count++;
-
-  // librdkafka API sometimes wants to write errors into a buffer:
-  std::vector<char> Errors;
-  Errors.resize(512);
-
-  rd_kafka_conf_t *Configuration = 0;
-  Configuration = rd_kafka_conf_new();
-  rd_kafka_conf_set_dr_msg_cb(Configuration, Producer::cb_delivered);
-  rd_kafka_conf_set_error_cb(Configuration, Producer::cb_error);
-  rd_kafka_conf_set_stats_cb(Configuration, Producer::cb_stats);
-  rd_kafka_conf_set_log_cb(Configuration, Producer::cb_log);
-  rd_kafka_conf_set_throttle_cb(Configuration, Producer::cb_throttle);
-
-  rd_kafka_conf_set_opaque(Configuration, this);
-  LOG(Sev::Debug, "Producer opaque: {}", (void *)this);
+  ProducerID = ProducerInstanceCount++;
 
-  ProducerBrokerSettings.apply(Configuration);
+  std::string ErrorString;
 
-  RdKafkaPtr = rd_kafka_new(RD_KAFKA_PRODUCER, Configuration, Errors.data(),
-                            Errors.size());
-  if (!RdKafkaPtr) {
-    LOG(Sev::Error, "can not create kafka handle: {}", Errors.data());
-    throw std::runtime_error("can not create Kafka handle");
+  Conf = std::unique_ptr<RdKafka::Conf>(
+      RdKafka::Conf::create(RdKafka::Conf::CONF_GLOBAL));
+  try {
+    ProducerBrokerSettings.apply(Conf.get());
+  } catch (std::runtime_error &e) {
+    throw std::runtime_error(
+        "Cannot create kafka handle due to configuration error");
   }
 
-  rd_kafka_set_log_level(RdKafkaPtr, 4);
-
-  LOG(Sev::Info, "New Kafka {} with brokers: {}", rd_kafka_name(RdKafkaPtr),
-      ProducerBrokerSettings.Address.c_str());
-  if (rd_kafka_brokers_add(RdKafkaPtr,
-                           ProducerBrokerSettings.Address.c_str()) == 0) {
-    LOG(Sev::Error, "could not add brokers");
-    throw std::runtime_error("could not add brokers");
+  Conf->set("dr_cb", &DeliveryCb, ErrorString);
+  Conf->set("event_cb", &EventCb, ErrorString);
+  Conf->set("metadata.broker.list", ProducerBrokerSettings.Address,
+            ErrorString);
+  ProducerPtr = std::unique_ptr<RdKafka::Producer>(
+      RdKafka::Producer::create(Conf.get(), ErrorString));
+  if (ProducerPtr == nullptr) {
+    LOG(Sev::Error, "can not create kafka handle: {}", ErrorString);
+    throw std::runtime_error("can not create Kafka handle");
   }
-}
 
-Producer::Producer(Producer &&x) {
-  using std::swap;
-  swap(RdKafkaPtr, x.RdKafkaPtr);
-  swap(on_delivery_ok, x.on_delivery_ok);
-  swap(on_delivery_failed, x.on_delivery_failed);
-  swap(on_error, x.on_error);
-  swap(ProducerBrokerSettings, x.ProducerBrokerSettings);
-  swap(id, x.id);
+  LOG(Sev::Info, "new Kafka producer: {}, with brokers: {}",
+      ProducerPtr->name(), ProducerBrokerSettings.Address.c_str());
 }
 
 void Producer::poll() {
-  int EventsHandled =
-      rd_kafka_poll(RdKafkaPtr, ProducerBrokerSettings.PollTimeoutMS);
+  auto EventsHandled = ProducerPtr->poll(ProducerBrokerSettings.PollTimeoutMS);
```

Comment:
Is poll actually used/needed?

Reply:
it's used for the kafkaGelf stuff so we should probably leave it